File: rules/rule_engine.py
class RuleEngine:

    # These are prototype/demo thresholds.
    # They should be replaced with evidence-backed values later.
    SUGAR_CAUTION = 15
    SODIUM_CAUTION = 400
    SATURATED_FAT_CAUTION = 5

    def assess(self, patient, verified_food, medicine_result=None):
        conditions = [
            condition.lower()
            for condition in patient.get("conditions", [])
        ]

        restrictions = [
            restriction.lower()
            for restriction in patient.get("dietary_restrictions", [])
        ]

        nutrition = verified_food.get("verified_nutrition", {})
        ingredients = [
            ingredient.lower()
            for ingredient in verified_food.get("ingredients", [])
        ]

        concerns = []
        rules_applied = []
        if medicine_result:
            possible_interactions = medicine_result.get(
                "possible_interactions", []
            )

            if possible_interactions:
                rules_applied.append("medicine_food_interaction_rule")

                for interaction in possible_interactions:
                    concerns.append({
                        "type": "medicine_food_interaction",
                        "medicine": interaction["medicine"],
                        "food_component": interaction["food_component"],
                        "message": interaction["message"]
                    })

        # --------------------------------
        # Diabetes / Low Sugar
        # --------------------------------

        sugar = nutrition.get("sugar")

        if (
            ("diabetes" in conditions or "low sugar" in restrictions)
            and sugar is not None
        ):

            rules_applied.append("diabetes_sugar_rule")

            if sugar >= self.SUGAR_CAUTION:

                concerns.append({
                    "type": "sugar",
                    "value": sugar,
                    "message": (
                        f"Sugar is {sugar} g per serving, "
                        "which exceeds the prototype caution threshold."
                    )
                })

        # --------------------------------
        # Hypertension / Low Sodium
        # --------------------------------

        sodium = nutrition.get("sodium")

        if (
            ("hypertension" in conditions or "low sodium" in restrictions)
            and sodium is not None
        ):

            rules_applied.append("hypertension_sodium_rule")

            if sodium >= self.SODIUM_CAUTION:

                concerns.append({
                    "type": "sodium",
                    "value": sodium,
                    "message": (
                        f"Sodium is {sodium} mg per serving, "
                        "which exceeds the prototype caution threshold."
                    )
                })

        # --------------------------------
        # Cardiovascular considerations
        # --------------------------------

        saturated_fat = nutrition.get("saturated_fat")

        if "cardiovascular" in conditions:

            rules_applied.append("cardiovascular_saturated_fat_rule")

            if (
                saturated_fat is not None
                and saturated_fat >= self.SATURATED_FAT_CAUTION
            ):

                concerns.append({
                    "type": "saturated_fat",
                    "value": saturated_fat,
                    "message": (
                        f"Saturated fat is {saturated_fat} g per serving, "
                        "which exceeds the prototype caution threshold."
                    )
                })

        # --------------------------------
        # Final Decision
        # --------------------------------

        if len(concerns) >= 2:

            decision = "HIGHER_CONCERN"

        elif len(concerns) == 1:

            decision = "CONSUME_WITH_CAUTION"

        else:

            decision = "GENERALLY_SUITABLE"

        return {
            "decision": decision,
            "concerns": concerns,
            "rules_applied": rules_applied
        }

File: src/agent/rule-engine.ts
import { MedicineResult } from './medicine-context';

export interface Concern {
  type: string;
  value?: number;
  medicine?: string;
  foodComponent?: string;
  message: string;
}

export interface RuleResult {
  decision: 'GENERALLY_SUITABLE' | 'CONSUME_WITH_CAUTION' | 'HIGHER_CONCERN';
  concerns: Concern[];
  rulesApplied: string[];
}

// Prototype thresholds — should be replaced with evidence-backed values later
const SUGAR_CAUTION = 15;
const SODIUM_CAUTION = 400;
const SATURATED_FAT_CAUTION = 5;

interface PatientInput {
  conditions?: string[];
  dietary_restrictions?: string[];
}

interface VerifiedFoodInput {
  verifiedNutrition?: Record<string, number>;
  ingredients?: string[];
}

export function assess(
  patient: PatientInput,
  verifiedFood: VerifiedFoodInput,
  medicineResult?: MedicineResult | null,
): RuleResult {
  const conditions = (patient.conditions || []).map((c) => c.toLowerCase());
  const restrictions = (patient.dietary_restrictions || []).map((r) => r.toLowerCase());
  const nutrition = verifiedFood.verifiedNutrition || {};

  const concerns: Concern[] = [];
  const rulesApplied: string[] = [];

  // Medicine-food interaction rules
  if (medicineResult?.possibleInteractions?.length) {
    rulesApplied.push('medicine_food_interaction_rule');
    for (const interaction of medicineResult.possibleInteractions) {
      concerns.push({
        type: 'medicine_food_interaction',
        medicine: interaction.medicine,
        foodComponent: interaction.foodComponent,
        message: interaction.message,
      });
    }
  }

  // Diabetes / Low Sugar
  const sugar = nutrition.sugar;
  if (
    (conditions.includes('diabetes') || restrictions.includes('low sugar')) &&
    sugar !== undefined
  ) {
    rulesApplied.push('diabetes_sugar_rule');
    if (sugar >= SUGAR_CAUTION) {
      concerns.push({
        type: 'sugar',
        value: sugar,
        message: `Sugar is ${sugar} g per serving, which exceeds the prototype caution threshold.`,
      });
    }
  }

  // Hypertension / Low Sodium
  const sodium = nutrition.sodium;
  if (
    (conditions.includes('hypertension') || restrictions.includes('low sodium')) &&
    sodium !== undefined
  ) {
    rulesApplied.push('hypertension_sodium_rule');
    if (sodium >= SODIUM_CAUTION) {
      concerns.push({
        type: 'sodium',
        value: sodium,
        message: `Sodium is ${sodium} mg per serving, which exceeds the prototype caution threshold.`,
      });
    }
  }

  // Cardiovascular / Saturated fat
  const saturatedFat = nutrition.saturated_fat;
  if (conditions.includes('cardiovascular')) {
    rulesApplied.push('cardiovascular_saturated_fat_rule');
    if (saturatedFat !== undefined && saturatedFat >= SATURATED_FAT_CAUTION) {
      concerns.push({
        type: 'saturated_fat',
        value: saturatedFat,
        message: `Saturated fat is ${saturatedFat} g per serving, which exceeds the prototype caution threshold.`,
      });
    }
  }

  // Final decision
  let decision: RuleResult['decision'];
  if (concerns.length >= 2) {
    decision = 'HIGHER_CONCERN';
  } else if (concerns.length === 1) {
    decision = 'CONSUME_WITH_CAUTION';
  } else {
    decision = 'GENERALLY_SUITABLE';
  }

  return { decision, concerns, rulesApplied };
}
